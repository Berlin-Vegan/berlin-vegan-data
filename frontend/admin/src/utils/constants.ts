export const GASTRO: string = 'GASTRO';
export const SHOPPING: string = 'SHOPPING';

export type LocationType = typeof GASTRO | typeof SHOPPING;

const API = '/api/';
export const V1 = `${API}v1/`;
export const V2 = `${API}v2/`;
