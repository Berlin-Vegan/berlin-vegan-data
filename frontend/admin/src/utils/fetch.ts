import { getCSRFToken } from './cookie';
import { UserDispatch, TYPE_USER_LOGOUT } from '../providers/UserProvider';

export const authorizedFetch = async (
  userDispatch: UserDispatch,
  url: string,
  method: string = 'GET',
  body: object | Array<unknown> | null = null
): Promise<Response> => {
  const response = await fetch(url, {
    method,
    body: body ? JSON.stringify(body) : body,
    headers: {
      'Content-Type': 'application/json',
      'X-CSRFToken': getCSRFToken(),
    },
  });

  if (response.status === 403) {
    userDispatch({ type: TYPE_USER_LOGOUT });
  }

  return response;
};

export const fetchGastroList = async (
  userDispatch: UserDispatch,
  filter: string
): Promise<Response> =>
  authorizedFetch(userDispatch, `/api/v1/gastros/?${filter}`);
