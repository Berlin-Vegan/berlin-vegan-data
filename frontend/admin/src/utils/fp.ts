import {
  addIndex,
  compose,
  curry,
  defaultTo,
  join,
  map,
  nth,
  pipe,
  splitAt,
  toLower,
  toUpper,
} from 'ramda';

export const nthOr = curry((defaultVal, list) =>
  compose(defaultTo(defaultVal), nth)(list),
);

export const nthOrEmpty = nthOr('');

export const mapIndexed = addIndex(map);

export const capitalizeWord = pipe(
  toLower,
  splitAt(1),
  (splitted) => [toUpper(nthOrEmpty(0)(splitted)), nthOrEmpty(1)(splitted)],
  join(''),
);
