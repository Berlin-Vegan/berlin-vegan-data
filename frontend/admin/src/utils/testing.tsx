import renderer from 'react-test-renderer';
import React, { ComponentType } from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';

export const standardComponentTest = (C: ComponentType<any>, props = {}) => {
  renderWithTestRenderer(C, props);
  it('renders without crashing', () => renders(C, props));
};

export const standardComponentTestMockFetch = (
  C: ComponentType<any>,
  props = {},
  fakeResponse: any,
) => {
  renderWithTestRenderer(C, props);
  it('renders without crashing and mock fetch', async () => {
    jest.spyOn(window, 'fetch').mockImplementation((): Promise<any> => {
      const fetchResponse = {
        json: () => Promise.resolve(fakeResponse),
      };
      return Promise.resolve(fetchResponse);
    });
    await act(async () => renders(C, props));
  });
};

export const renders = (C: ComponentType, props: object) => {
  const div = document.createElement('div');
  ReactDOM.render(<C {...props} />, div);
};

const getName = (C: ComponentType) => C.displayName || C.name || 'A Component';

const renderWithTestRenderer = (C: ComponentType, props: object) => {
  it(`render <${getName(C)} />`, () => {
    const tree = renderer.create(<C {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
};
