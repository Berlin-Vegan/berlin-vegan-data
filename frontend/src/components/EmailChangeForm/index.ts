import EmailChangeForm from './component';

export default EmailChangeForm;
