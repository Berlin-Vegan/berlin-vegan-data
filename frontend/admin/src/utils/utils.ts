import { pipe, toLower } from 'ramda';
import { LocationType, V1, V2 } from './constants';

//
export const buildReviewListUrl: string = `${V1}review/`;
export const buildReviewDetailUrl = (id: number): string =>
  `${buildReviewListUrl}${id}/`;

// API V2 urls
export const buildListUrl = (type: LocationType): string =>
  pipe(toLower, (lowerType: string) => `${V2}${lowerType}/`)(type);
export const buildDetailUrl = (type: LocationType, id: string): string =>
  pipe(toLower, buildListUrl, (baseUrl: string) => `${baseUrl}${id}/`)(type);
// FE urls
export const buildFEDetailUrl = (type: LocationType, id: string): string =>
  pipe(toLower, (lowerType: string) => `/${lowerType}/${id}/edit`)(type);
