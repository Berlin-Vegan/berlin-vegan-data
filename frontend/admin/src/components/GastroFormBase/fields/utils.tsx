import React from 'react';
import MenuItem from '@material-ui/core/MenuItem';
import { map } from 'ramda';
import { nthOr } from '../../../utils/fp';

const menuItem = (value: number | string, label: string) => (
  <MenuItem value={value} key={value}>
    {label}
  </MenuItem>
);

export const menuItems = (fieldOptions: Array<[number | string, string]>) =>
  map((item) => menuItem(nthOr('', 0)(item), nthOr('', 1)(item)))(fieldOptions);

export const withLeadingZero = (value: number): string =>
  value < 10 ? `0${value}` : `${value}`;
