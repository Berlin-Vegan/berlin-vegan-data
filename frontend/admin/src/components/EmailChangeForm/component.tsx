import React, { useContext } from 'react';
import { useSnackbar } from 'notistack';
import { Field, Form, Formik } from 'formik';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import { TextField } from 'formik-material-ui';
import { Button } from '@material-ui/core';
import SaveIcon from '@material-ui/icons/Save';
import { makeStyles } from '@material-ui/core/styles';
import * as Yup from 'yup';
import { pathOr } from 'ramda';
import { authorizedFetch } from '../../utils/fetch';
import {
  AuthContext,
  TYPE_UPDATE_USER_EMAIL,
} from '../../providers/UserProvider';

const useStyles = makeStyles(() => ({
  saveButton: {
    marginTop: '10px',
  },
}));

const emailChangeForm = Yup.object().shape({
  email: Yup.string().email().nullable(),
});

const profileURL = '/api/v1/accounts/profile/';

const EmailChangeForm = () => {
  const classes = useStyles();
  const { enqueueSnackbar } = useSnackbar();
  const { state, dispatch } = useContext(AuthContext);
  const email = pathOr('', ['userData', 'email'], state);

  return (
    <Formik
      initialValues={{ email }}
      validationSchema={emailChangeForm}
      onSubmit={async (values, { setSubmitting, setErrors }) => {
        const res = await authorizedFetch(
          dispatch,
          profileURL,
          'PATCH',
          values,
        );
        const data = await res.json();
        if (res.status === 200) {
          enqueueSnackbar('Email updated', {
            variant: 'success',
          });
          dispatch({ type: TYPE_UPDATE_USER_EMAIL, payload: values.email });
        }
        if (res.status === 400) {
          enqueueSnackbar('Form invalid', {
            variant: 'error',
          });
          setErrors(data);
        }
        setSubmitting(false);
      }}
    >
      {({ submitForm, isSubmitting }) => (
        <Form>
          <Grid container spacing={1} direction="column">
            <Grid container item direction="column" md={4}>
              <Typography variant="h5">Change Email</Typography>
              <Field
                component={TextField}
                name="email"
                type="email"
                label="E-Mail"
                required
              />
              <Grid container justify="flex-end" className={classes.saveButton}>
                <Button
                  variant="contained"
                  color="primary"
                  disabled={isSubmitting}
                  onClick={submitForm}
                  name="save"
                  startIcon={<SaveIcon />}
                >
                  Save
                </Button>
              </Grid>
            </Grid>
          </Grid>
        </Form>
      )}
    </Formik>
  );
};

export default EmailChangeForm;
