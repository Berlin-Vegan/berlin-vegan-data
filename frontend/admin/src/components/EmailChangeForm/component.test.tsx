import React from 'react';
import { SnackbarProvider } from 'notistack';
import { standardComponentTest } from '../../utils/testing';
import EmailChangeForm from './component';

const EmailChangeFormWrapper = (initial: object) => (
  <SnackbarProvider maxSnack={3}>
    <EmailChangeForm />
  </SnackbarProvider>
);

standardComponentTest(EmailChangeFormWrapper);
