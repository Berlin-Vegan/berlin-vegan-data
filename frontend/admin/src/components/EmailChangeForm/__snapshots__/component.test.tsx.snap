// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render <EmailChangeFormWrapper /> 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-direction-xs-column"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-md-4"
    >
      <h5
        className="MuiTypography-root MuiTypography-h5"
      >
        Change Email
      </h5>
      <div
        className="MuiFormControl-root MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
          data-shrink={false}
        >
          E-Mail
          <span
            aria-hidden={true}
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
            â€‰
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            name="email"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="email"
            value=""
          />
        </div>
      </div>
      <div
        className="MuiGrid-root makeStyles-saveButton-1 MuiGrid-container MuiGrid-justify-xs-flex-end"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          disabled={false}
          name="save"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            <span
              className="MuiButton-startIcon MuiButton-iconSizeMedium"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
                />
              </svg>
            </span>
            Save
          </span>
        </button>
      </div>
    </div>
  </div>
</form>
`;
